enhancements/results-forwarding.md

```diff
@@ -0,0 +1,504 @@
+
+A user may explicitly disable in-cluster storage if forwarding is enabled. The
+Compliance Operator will not support disabling PVC storage without a valid
+forwarding endpoint.
+
+This will require changing the **aggregator** to always forward via a gRPC API
+to a configured implementation. In this case, the **aggregator** will act as a
+gRPC client. The **resultserver** then becomes a gRPC server, subject to the
+evidence forwarding provider and will be renamed **evidence-persistor** and
+responsible for writing results.
+
+Forwarder implementations would be configured via the `ScanSetting` and they'd be called "providers".
+For backwards compatibility, we'd have defaults that would point to the current mode of operation
+of the Compliance Operator. e.g. a `pvc` provider for evidence storage.
+
+For more detail on the proposed API, see the *API Extensions* section.
+
+### User Stories
+
+* As an Site Reliability Engineer managing multiple clusters, I'd like to run
+  compliance scans on my fleet in such a way that uses the minimal amount of
+  resources.
+
+* As a Site Reliability Engineer managing multiple clusters, I'd like to have
+  one place to view the compliance stance of my whole fleet.
+
+### API Extensions
+
+Given the scope of the change, we'd upgrade the `ScanSetting` version to
+`v1alpha2`. The following sections describe the additional values needed by the
+`ScanSetting` CRD.
+
+#### `resultForwarding.provider`
+
+Required: True
+Type: string
+
+The implementation to use for forwarding. The initial implementation will
+support `grpc`.
+
+#### `resultForwarding.grpc.endpoint`
+
+Required: True
+Type: string
+
+Endpoint URL to a gRPC server for result storage. The service may be in or
+outside the cluster. One potential implementation may be
+[compserv](https://github.com/rhmdnd/compserv).
+
+#### `resultForwarding.grpc.version`
+
+Required: False
+Default: `v1`
+Type: string
+
+#### `resultForwarding.grpc.tls.dynamic.caSecretName`
+
+Required: True
+Type: string
+
+Reference to a CA certificate and key pair for TLS.
+
+#### `resultForwarding.grpc.tls.dynamic.serverSecretName`
+
+Required: True
+Type: string
+
+Reference to a server certificate/keypair secret for mutual TLS.
+
+#### `resultForwarding.grpc.tls.dynamic.clientSecretName`
+
+Required: True if using mutual TLS
+Type: string
+
+Reference to a client certificate/keypair secret for mutual TLS.
+
+#### `resultForwarding.grpc.authentication.type`
+
+Required: False
+Default: `mtls`
+Type: string
+Choices: `mtls`, `token`
+
+The type of authentication the Compliance Operator should use to authenticate
+the forwarding endpoint.
+
+#### `resultForwarding.grpc.authentication.token`
+
+Required: True if using `token` authentication type
+Type: string
+
+#### `resultForwarding.grpc.authentication.token.tokenSecretName`
+
+Required: False
+Type: string
+
+Reference to the secret containing the token.
+
+#### `resultForwarding.grpc.extraMetadata.clusterName`
+
+Required: True
```

Comment:
If this information is required, and we do agree that it must be specified in the `ScanSetting`, I wonder if we should treat it more formally than `extra`.

`extraMetadata` feels like a catch all, whereas the `clusterName` is certainly something we need to send to the external endpoint.

Reply:
Maybe we can use `resultForwarding.grpc.metadata.clusterName` instead?

Reply:
Yeah - or even making it a first-class attribute of the `grpc` map (e.g., `resultForwarding.grpc.clusterName`)?

Reply:
that depends on the use-case. Are we gonna allow a single server to send results via GRPC? in that case we wouldn't quite need the cluster name and instead we would rely on other information. e.g. availability zone, geographical location, etc. This depends on where else do you want to take this result forwarding capability.

Reply:
> that depends on the use-case. Are we gonna allow a single server to send results via GRPC?

Do you think this attribute should be generalized to just a name? Also, can you provide an example of what you mean by "single server"?

Reply:
I'd keep `extraMetadata` as a dictionary that operators can build themselves and provide examples and conventions on how to use it. One case can be a kubernetes cluster, where you're interested in knowing the cluster name, node pool and perhaps even region.

From a central result perspective, it could be the case that to compserv you allow folks to push results from non-cluster entities. e.g. a single RHEL node running a legacy service. From that kind of entity, you'll probably need different information.